arai60-347/solution.md

```diff
@@ -0,0 +1,105 @@
+# 1st
+```python
+from collections import defaultdict
+import heapq
+class Solution:
+    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
+        num_frequency = defaultdict(int)
+        top_frequent_nums = []
+        top_k_nums = []
+        for num in nums:
+            num_frequency[num] += 1
+        for num in num_frequency:
+            heapq.heappush(top_frequent_nums, (-num_frequency[num], num))
+        for i in range(k):
+            _, num = heapq.heappop(top_frequent_nums)
+            top_k_nums.append(num)
+        return top_k_nums
+```
+# 2nd 
+- https://github.com/YukiMichishita/LeetCode/pull/3/files#diff-1897b7a4643771fb323fe5359b1c63a3cf56fec3f14f3271974b6c5cc9231340
+    - `for num, frequency in nums_to_count.items():`ってできるの知らなかった
+- https://github.com/YukiMichishita/LeetCode/pull/3/files#diff-36345f1a68047c1e6115ff93e039240db986e0f001b2b7aa192613739057a8e7
+  - quick selectでも解ける
+```py
+from collections import defaultdict
+import heapq
+class Solution:
+    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
+        num_to_count = defaultdict(int)
+        num_frequency_heap = []
+        top_k_nums = []
+        for num in nums:
+            num_to_count[num] += 1
+        for num , count in num_to_count.items():
+            heapq.heappush(num_frequency_heap, (-count, num))
+        for _ in range(k):
+            _, num = heapq.heappop(num_frequency_heap)
+            top_k_nums.append(num)
+        return top_k_nums
+```
+
+# 3rd
+```py
+from collections import defaultdict
+import heapq
+class Solution:
+    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
+        num_to_count = defaultdict(int)
+        num_frequency_heap = []
+        top_k_frequent_nums = []
+        for num in nums:
+            num_to_count[num] += 1
+        for num, count in num_to_count.items():
+            heapq.heappush(num_frequency_heap, (-count, num))
+        for _ in range(k):
+            _, num = heapq.heappop(num_frequency_heap)
+            top_k_frequent_nums.append(num)
+        return top_k_frequent_nums
+```
+
+# 4th
+- 変数名に(-count, num)が入っていることを説明するようにした。
+- 変数宣言は実際に使う直前で宣言するように修正した。
+```py
+from collections import defaultdict
+import heapq
+class Solution:
+    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
+        num_to_count = defaultdict(int)
+        for num in nums:
+            num_to_count[num] += 1
+        minus_count_and_num_heap = []
+        for num, count in num_to_count.items():
+            heapq.heappush(minus_count_and_num_heap, (-count, num))
+        top_k_frequent_nums = []
+        for _ in range(k):
+            _, num = heapq.heappop(minus_count_and_num_heap)
+            top_k_frequent_nums.append(num)
+        return top_k_frequent_nums
+```
+
+# defaultdictの実装
+- クラスのメンバにdictを持たせて実装
+- new_dict[3]はnew_dict.__getitem__(3)のsyntax sugar
+```py
+class my_default_dict:
```

Comment:
https://peps.python.org/pep-0008/#class-names
Class names should normally use the CapWords convention.